skcosmo/preprocessing/FlexibleScaler.py

```diff
@@ -0,0 +1,242 @@
+import numpy as np
+import sklearn as sk
+from sklearn.base import TransformerMixin, BaseEstimator
+
+
+class StandardFlexibleScaler(TransformerMixin, BaseEstimator):
+    """A more flexible scaler for features and data, that mimics
+    sklearn.preprocessing.StandardScaler make the mean of the
+    columns equal to zero and the variance of each column
+    (if designated) equal to one.
+
+    :param with_mean: If True, center the data before scaling
+    :type with_mean: boolean
+    :param with_std: If True, scale the data to unit variance
+    :type with_std: boolean
+    :param column_wise: If True, normalize each column separately
+    :type column_wise: boolean
+    """
+
+    def __init__(self, *, with_mean=True, with_std=True, column_wise=False):
+        """Initialize StandardFlexibleScaler.
+
+
+        :param with_mean: If True, center the data before scaling
+        :type with_mean: boolean
+        :param with_std: If True, scale the data to unit variance
+        :type with_std: boolean
+        :param column_wise: If True, normalize each column separately
+        :type column_wise: boolean
+        """
+
+        self.with_mean = with_mean
+        self.with_std = with_std
+        self.column_wise = column_wise
+        self.n_samples_seen_ = 0
+
+    def fit(self, X, y=None):
+        """Compute mean and scaling to be applied for subsequent normalization.
+
+        :param X: Matrix
+        :type X: ndarray
+        :param y: ignored
+
+        :return: itself
+        """
+
+        self.n_samples_seen_, self.n_features_ = X.shape
+        self.mean_ = np.zeros(self.n_features_)
+
+        if self.with_mean:
+            self.mean_ = X.mean(axis=0)
+            centred_X = X - self.mean_
+        else:
+            centred_X = X
+
+        self.scale_ = 1.0
+        if self.with_std:
+            var = ((centred_X - centred_X.mean(axis=0)) ** 2).mean(axis=0)
+
+            if self.column_wise:
+                if np.any(var == 0):
+                    raise ValueError("Cannot normalize a feature with zero variance")
+                self.scale_ = np.sqrt(var)
+            else:
+                var_sum = var.sum()
+                if var_sum == 0:
+                    raise ValueError("Cannot normalize a matrix with zero variance")
+                self.scale_ = np.sqrt(var_sum)
```

Comment:
Then here we can use X directly to determine the scale. Otherwise just some code simplification
```suggestion
        self.scale_ = 1.0
        if self.with_std:
            # we always compute the variance of the centered X, 
            # regardless of whether we intend to re-center it.
            scale_sq = ((X - X.mean(axis=0)) ** 2).mean(axis=0)

            if not self.column_wise:
                scale_sq = scale_sq.sum()
                
            if np.any(scale_sq == 0):
                raise ValueError("Cannot normalize a feature or matrix with zero variance")
                
            self.scale_ = np.sqrt(scale_sq)

```

Reply:
I disagree a little with this. I would be confused by such an error comment. For example, I will have a matrix with non-zero variation and 1 column with zero variation. And I will not understand what exactly is the reason for the error. In my opinion, it is important to keep these two options separate.

Reply:
How about this? You should also use a `< some_theshold` instead of `==0`, as there may be some near-zero values that may cause a problem.

```suggestion
        self.scale_ = 1.0
        if self.with_std:
            # we always compute the variance of the centered X, 
            # regardless of whether we intend to re-center it.
            scale_sq = ((X - X.mean(axis=0)) ** 2).mean(axis=0)
            
            if not self.column_wise:
                scale_sq = scale_sq.sum()
                if np.any(scale_sq == 0):
                    raise ValueError("Cannot normalize a matrix with zero variance")
            elif np.any(scale_sq == 0):
                zero_feat = np.where(scale_sq == 0)[0]
                raise ValueError("Cannot normalize a feature with zero variance; feature[s] {zero_feat} have zero variance.")
                
            self.scale_ = np.sqrt(scale_sq)
```

Reply:
Ok, yes, that sounds reasonable